src/adapter/src/coord.rs

```diff
@@ -363,13 +367,30 @@ pub enum TargetCluster {
     Active,
 }
 
-/// A struct to hold information on how to determine if a plan may have changed and how to re-plan
-/// it if so.
+/// A struct to hold information about the validity of peeks and if they should be abandoned after
+/// doing work off of the Coordinator thread.
 #[derive(Debug)]
-pub struct TransientPlan {
+pub struct PeekValidity {
+    /// The most recent revision at which this peek was verified as valid.
     transient_revision: u64,
-    plan: mz_sql::plan::Plan,
-    depends_on: Vec<GlobalId>,
+    /// Objects on which the peek depends.
+    source_ids: BTreeSet<GlobalId>,
+}
+
+impl PeekValidity {
+    /// Returns an error if the current catalog no longer has all dependent sources.
+    fn check(&mut self, catalog: &Catalog) -> Result<(), AdapterError> {
+        if self.transient_revision == catalog.transient_revision() {
+            return Ok(());
+        }
+        for id in &self.source_ids {
+            if catalog.try_get_entry(id).is_none() {
```

Comment:
I'm like *pretty??* sure this is correct because:

- we never mutate ids while running
- if an id was dropped we'll detect it here

Reply:
I tried to do this exact thing in my RTR PR and I thought you had a good point about why this might be a bad idea: https://github.com/MaterializeInc/materialize/pull/17127/files/d9cc505a0b6e95b13305e85325e96d74cc4068c8#r1069680520

Reply:
We should never be re-using IDs. This looks correct now, but I think the point of the comment I linked is that if something changes in the future, then this may silently become incorrect. 

For a concrete example, the `assignment` and `returning` fields of a `ReadThenWritePlan` plan are not accessible here. We have a current limitation that those two fields can only reference the table that is being mutated so it's not an issue. If we ever update read-then-write queries so that `assignment` and `returning` can be arbitrary queries (which is allowed in PostgreSQL), then that may break the assumption being made here.